burst-gradle-plugin/src/test/kotlin/app/cash/burst/gradle/TestInterceptorGradlePluginTest.kt

```diff
@@ -0,0 +1,405 @@
+      assertThat(systemOut).isEqualTo(
+        """
+        |intercepting
+        |  packageName=app.cash.burst.tests
+        |  className=BasicTest
+        |  functionName=passingTest
+        |set up
+        |running
+        |tear down
+        |intercepted
+        |
+        """.trimMargin(),
+      )
+    }
+  }
+
+  @Test
+  fun failingTest() {
+    with(tester.readTestSuite("app.cash.burst.tests.FailingTest")) {
+      assertThat(testCases.single().failureMessage).isNotNull().contains("boom!")
+      assertThat(systemOut).isEqualTo(
+        """
+        |set up
+        |tear down
+        |re-throwing exception: boom!
+        |
+        """.trimMargin(),
+      )
+    }
+  }
+
+  /**
+   * In JUnit 4:
+   *   multiple set up functions run in reverse alphabetical order
+   *   multiple tear down functions run in alphabetical order
+   */
+  @Test
+  fun multipleBeforesAndAfters() {
+    with(tester.readTestSuite("app.cash.burst.tests.MultipleBeforesAndAftersTest")) {
+      assertThat(testCases.single().failureMessage).isNull()
+      assertThat(systemOut).isEqualTo(
+        """
+        |intercepting
+        |beforeTest red
+        |beforeTest green
+        |beforeTest blue
+        |running
+        |afterTest blue
+        |afterTest green
+        |afterTest red
+        |intercepted
+        |
+        """.trimMargin(),
+      )
+    }
+  }
+
+  @Test
+  fun multipleAftersAfterFailure() {
+    with(tester.readTestSuite("app.cash.burst.tests.MultipleAftersAfterFailureTest")) {
+      assertThat(testCases.single().failureMessage).isNotNull().contains("boom!")
+      assertThat(systemOut).isEqualTo(
+        """
+        |afterTest blue
+        |afterTest green
+        |afterTest red
+        |re-throwing exception: boom!
+        |
+        """.trimMargin(),
+      )
+    }
+  }
+
+  /**
+   * If any @AfterTest throws, the other @AfterTest functions still execute.
+   */
+  @Test
+  fun afterTestThrows() {
+    with(tester.readTestSuite("app.cash.burst.tests.AfterTestThrowsTest")) {
+      assertThat(testCases.single().failureMessage).isNotNull().contains("boom!")
+      assertThat(systemOut).isEqualTo(
+        """
+        |running
+        |afterTest green
+        |afterTest red
+        |re-throwing exception: boom!
+        |
+        """.trimMargin(),
+      )
+    }
+  }
+
+  /**
+   * If any @BeforeTest throws, no more @BeforeTests run and neither does the test itself.
+   */
+  @Test
+  fun beforeTestThrows() {
+    with(tester.readTestSuite("app.cash.burst.tests.BeforeTestThrowsTest")) {
+      assertThat(testCases.single().failureMessage).isNotNull().contains("boom!")
+      assertThat(systemOut).isEqualTo(
+        """
+        |beforeTest red
+        |beforeTest green
+        |afterTest
+        |re-throwing exception: boom!
+        |
+        """.trimMargin(),
+      )
+    }
+  }
+
+  /**
+   * In JUnit 4, @Rules execute in reverse alphabetical order.
```

Comment:
Is this actually specified? Or is it an artifact of the implementation detail of JVM reflection order? The `@Rule` annotation does have an `order` integer (added in 4.13), but historically they recommended `RuleChain` to define a stable order. I don't recall ever seeing a declared order for multiple rules.

Reply:
Good call. I’ve fixed it to use declaration order.